compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5086,7 +5086,7 @@ private BType checkAndGetType(BLangExpression expr, SymbolEnv env, BLangBinaryEx
         this.dlog.mute();
 
         expr.cloneAttempt++;
-        BType exprCompatibleType = checkExpr(nodeCloner.cloneNode(expr), env, binaryExpr.expectedType, data);
+        BType exprCompatibleType = checkExpr(nodeCloner.cloneNodeOnce(expr), env, binaryExpr.expectedType, data);
```

Comment:
Do you have a narrowed down sample for which the issue is reproducible?

The whole point of cloning and type-checking is to allow checking a node against different types. So we need a new clone. I'm not sure if this fix is correct.

Reply:
> Do you have a narrowed down sample for which the issue is reproducible?
> 

This is a narrowed example. 
https://github.com/wso2-enterprise/internal-support-ballerina/issues/344#issuecomment-1579917157
 In this example, we create cloned nodes for lhs exponentially.

Reply:
Do we know why that happens? If it is cloned only as intended, shouldn't cause an issue, right?

Reply:
Don't we have to clean other usages of this? Add a test case, given in the example.

Reply:
Had a chat with Kavindu regarding this.




> The whole point of cloning and type-checking is to allow checking a node against different types. So we need a new clone.

For something like `A + B + C`, which is effectively a binary expression with the operands `A + B` and `C`, we clone `A + B` first, but then also clone `A` and `B` separately, With this fix we try to avoid cloning `A` and `B` if `A + B` is cloned.  

But, I think the correct fix is to clone only if required. For example, we don't have to clone `A + B` here, we only need to clone `A` and `B` if they are literals/require inference. We may be able to reuse `org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker#requireTypeInference`, but we need to make sure that the type is propagated properly as required. 

This is also reproducible with a type cast expression.

```ballerina
public function main() {
    decimal totalOtherCost = <decimal>  
                             <decimal>  
                             <decimal> 
                             <decimal>  
                             <decimal> 
                             <decimal> 
                             <decimal> 
                             <decimal> 
                             <decimal> 
                             <decimal> 
                             <decimal> 
                             <decimal> 
                             <decimal>  
                             <decimal> 
                             <decimal>  
                             <decimal> 
                             <decimal>  
                             <decimal> 
                             <decimal>  
                             <decimal> 
                             <decimal>  
                             <decimal> 
                             <decimal>  
                             <decimal> 1;
}
```

Reply:
Btw, we did previously use `requireTypeInference` for type casts, but removed that check later in https://github.com/ballerina-platform/ballerina-lang/pull/30655. We need to check similar scenarios with the fix we come up for this.

Reply:
ACK. Change the logic to call the node cloner only for numeric literal.

I will check the type-cast scenario and change it too.